Add json serialization to serialization overview doc (#16484)
dotnet/docs: docs/csharp/programming-guide/concepts/serialization/index.md

![Serialization graphic](./media/index/serialization-process.gif)

The object is serialized to a stream that carries the data. The stream may also have information about the object's type, such as its version, culture, and assembly name. From that stream, the object can be stored in a database, a file, or memory.

### Uses for serialization

Serialization allows the developer to save the state of an object and re-create it as needed, providing storage of objects as well as data exchange. Through serialization, a developer can perform actions such as:

* Sending the object to a remote application by using a web service
* Passing an object from one domain to another
* Passing an object through a firewall as a JSON or XML string
* Maintaining security or user-specific information across applications

## JSON serialization

> The [System.Text.Json namespace](../../../../standard/serialization/system-text-json-overview.md) contains classes for JavaScript Object Notation (JSON) serialization and deserialization. JSON is an open standard that is commonly used for sharing data across the web.

Reviewer: Shouldn't this link to the API ref instead of the overview doc?

Author: For this doc's audience -- people new to serialization -- it seems to me that the overview doc is more useful than the namespace API ref. The link text does make it look like you'd get an API ref doc, though, so I can see why that might be a good choice as well.

Reviewer: Other links point to the API ref. The namespace should point to the API imo. Maybe you can add a sentence on how to use it and have that link to the how-to doc.

> JSON serialization serializes the public properties of an object into a string, byte array, or stream that conforms to [the RFC 8259 JSON specification](https://tools.ietf.org/html/rfc8259#section-7). To control the way <xref:System.Text.Json.JsonSerializer> serializes or deserializes an instance of the class:

Reviewer: Why section 7?
```suggestion
JSON serialization serializes the public properties of an object into a string, byte array, or stream that conforms to [the RFC 8259 JSON specification](https://tools.ietf.org/html/rfc8259). To control the way <xref:System.Text.Json.JsonSerializer> serializes or deserializes an instance of the class:
```

* Use a [JsonSerializerOptions](xref:System.Text.Json.JsonSerializerOptions) object

> * Apply attributes from the [System.Text.Json](xref:System.Text.Json.Serialization) namespace to classes or properties

Reviewer: ```suggestion
* Apply attributes from the [System.Text.Json.Serialization](xref:System.Text.Json.Serialization) namespace to classes or properties
```

When you use basic serialization, the versioning of objects may create problems. You would use custom serialization when versioning issues are important. Basic serialization is the easiest way to perform serialization, but it does not provide much control over the process.

In custom serialization, you can specify exactly which objects will be serialized and how it will be done. The class must be marked <xref:System.SerializableAttribute> and implement the <xref:System.Runtime.Serialization.ISerializable> interface. If you want your object to be deserialized in a custom manner as well, use a custom constructor.

## Designer serialization

Designer serialization is a special form of serialization that involves the kind of object persistence associated with development tools. Designer serialization is the process of converting an object graph into a source file that can later be used to recover the object graph. A source file can contain code, markup, or even SQL table information.

## <a name="BKMK_RelatedTopics"></a> Related Topics and Examples  

> [System.Text.Json overview](../../../../standard/serialization/system-text-json-overview.md)

Reviewer: nit: Consider adding a one-sentence description for each of these additional resources like the existing ones below.

In custom serialization, you can specify exactly which objects will be serialized and how it will be done. The class must be marked <xref:System.SerializableAttribute> and implement the <xref:System.Runtime.Serialization.ISerializable> interface. If you want your object to be deserialized in a custom manner as well, use a custom constructor.

## Designer serialization

Designer serialization is a special form of serialization that involves the kind of object persistence associated with development tools. Designer serialization is the process of converting an object graph into a source file that can later be used to recover the object graph. A source file can contain code, markup, or even SQL table information.

## <a name="BKMK_RelatedTopics"></a> Related Topics and Examples  

[System.Text.Json overview](../../../../standard/serialization/system-text-json-overview.md)
Shows how to get the `System.Text.Json` library.

[How to serialize and deserialize JSON in .NET](../../../../standard/serialization/system-text-json-how-to.md). 

> Shows how to use the `System.Text.Json` library.

Reviewer: Following the existing descriptions:
```suggestion
Shows how to read and write object data to and from JSON using the <xref:System.Text.Json.JsonSerializer> class.
```
